File: app/src/main/java/br/com/popularmovies/data/ServiceGenerator.java
package br.com.popularmovies.data;

import com.facebook.stetho.okhttp3.StethoInterceptor;
import com.squareup.moshi.Moshi;

import br.com.popularmovies.BuildConfig;
import br.com.popularmovies.data.interceptor.AuthorizationInterceptor;
import br.com.popularmovies.utils.BigDecimalAdapter;
import br.com.popularmovies.utils.DateAdapter;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.moshi.MoshiConverterFactory;

import static br.com.popularmovies.data.Constants.API_DATE_PATTERN;
import static br.com.popularmovies.data.Constants.API_VERSION;
import static br.com.popularmovies.data.Constants.HOST;
import static br.com.popularmovies.data.Constants.SCHEME;

public class ServiceGenerator {


    private static final Moshi moshiFactory = new Moshi.Builder()
            .add(new BigDecimalAdapter())
            .add(new DateAdapter(API_DATE_PATTERN))
            .build();

    private static Retrofit.Builder builder =
            new Retrofit.Builder()
                    .baseUrl(buildUrl())
                    .addConverterFactory(MoshiConverterFactory.create(moshiFactory));

    private static Retrofit retrofit = builder.build();

    private static HttpLoggingInterceptor logging =
            new HttpLoggingInterceptor()
                    .setLevel(HttpLoggingInterceptor.Level.BODY);

    private static OkHttpClient.Builder httpClient =
            new OkHttpClient.Builder()
                    .addInterceptor(new AuthorizationInterceptor());

    public static <S> S createService(
            Class<S> serviceClass) {
        if (!httpClient.interceptors().contains(logging) && BuildConfig.DEBUG) {
            httpClient.addNetworkInterceptor(new StethoInterceptor());
            httpClient.addInterceptor(logging);
            builder.client(httpClient.build());
            retrofit = builder.build();
        }

        return retrofit.create(serviceClass);
    }

    private static HttpUrl buildUrl() {
        return new HttpUrl.Builder()
                .scheme(SCHEME)
                .host(HOST)
                .addPathSegments(API_VERSION)
                .build();
    }
}

File: app/src/main/java/br/com/popularmovies/core/network/retrofit/ServiceGenerator.kt
package br.com.popularmovies.core.network.retrofit

import br.com.popularmovies.BuildConfig
import br.com.popularmovies.core.network.API_VERSION
import br.com.popularmovies.core.network.HOST
import br.com.popularmovies.core.network.SCHEME
import br.com.popularmovies.core.network.retrofit.interceptor.AuthorizationInterceptor
import br.com.popularmovies.utils.BigDecimalAdapter
import br.com.popularmovies.utils.DateAdapter
import com.facebook.stetho.okhttp3.StethoInterceptor
import com.squareup.moshi.Moshi
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

object ServiceGenerator {

    private val moshiFactory = Moshi.Builder()
        .add(BigDecimalAdapter())
        .add(DateAdapter())
        .build()

    private val builder = Retrofit.Builder()
        .baseUrl(buildUrl())
        .addConverterFactory(MoshiConverterFactory.create(moshiFactory))

    private var retrofit = builder.build()

    private val logging = HttpLoggingInterceptor()
        .setLevel(HttpLoggingInterceptor.Level.BODY)

    private val httpClient = OkHttpClient.Builder()
        .addInterceptor(AuthorizationInterceptor())

    fun <S> createService(
        serviceClass: Class<S>
    ): S {
        if (!httpClient.interceptors().contains(
                logging
            ) && BuildConfig.DEBUG) {
            httpClient.addNetworkInterceptor(StethoInterceptor())
            httpClient.addInterceptor(logging)
            builder.client(httpClient.build())
            retrofit = builder.build()
        }

        return retrofit.create(serviceClass)
    }

    private fun buildUrl(): HttpUrl {
        return HttpUrl.Builder()
            .scheme(SCHEME)
            .host(HOST)
            .addPathSegments(API_VERSION)
            .build()
    }
}
